multipass-addressing/src/main/scala/com/precisely/bigdata/sample/spark/MultipassAddressing.scala

```diff
@@ -0,0 +1,159 @@
+    // of the specified output fields as well as an error column
+    addressInputDF
+      // Adds a new column, represented as a collection comprised of the outputFields and the error field
+      .withColumn("addressing_geocode_result", addressingUdf(map(
+        lit("street"), col("streetName"),
+        lit("city"), col("city"),
+        lit("admin1"), col("state"),
+        lit("postalCode"), col("zip")))
+      )
+      // Persist the addressing result to avoid recalculation when we expand the result
+      .persist()
+      // Expand the result collection such that each output field is a separate column, including the error field.
+      .select("*", "addressing_geocode_result.*").drop("addressing_geocode_result")
+      // Write the dataframe to the specified output folder as parquet file
+      .write.mode(SaveMode.Overwrite).option("header","true").csv(outputDirectory)
+  }
+}
+
+/*
+ * A custom Executor that contains the logic to perform verify and then multipass geocode
+ */
+class CustomExecutor extends com.precisely.bigdata.addressing.spark.api.AddressingExecutor {
+
+  import com.precisely.addressing.v1.model.Response
+  import com.precisely.addressing.v1.{Addressing, Preferences}
+
+  var relaxedGeocoderPreferences: Preferences = _
+
+  override def execute(input: RequestInput, preferences: Option[Preferences], addressing: Addressing): Response = {
+    if (relaxedGeocoderPreferences == null) {
+      relaxedGeocoderPreferences = new PreferencesBuilder()
+        .withMatchMode("RELAXED")
+        .build()
+    }
+
+    val verifyResult = addressing.verify(input.requestAddress(), preferences.orNull).getResults
+    val verifyAddress = verifyResult.get(0).getAddress
+    
+    //  create multiLine request from the result fields obtained from verify api
+    val multiLineRequest: RequestAddress = new RequestAddress()
+    multiLineRequest.setStreet(verifyAddress.getStreet)
+    multiLineRequest.setAddressNumber(verifyAddress.getAddressNumber)
+    multiLineRequest.setAdmin1(verifyAddress.getAdmin1.getLongName)
+    multiLineRequest.setAdmin2(verifyAddress.getAdmin2.getLongName)
+    multiLineRequest.setCountry(verifyAddress.getCountry.getIsoAlpha3Code)
+    multiLineRequest.setCity(verifyAddress.getCity.getLongName)
+    multiLineRequest.setPostalCode(verifyAddress.getPostalCode)
+
+    val multiLineResponse = addressing.geocode(multiLineRequest, preferences.orNull)
+    val multiLineResult = multiLineResponse.getResults
+
+    if (CollectionUtils.isNotEmpty(multiLineResult) && multiLineResult.get(0).getCustomFields
+      .get("PRECISION_CODE") != null && Seq("S5", "S6", "S7", "S8").contains(multiLineResult
+      .get(0).getCustomFields.get("PRECISION_CODE").slice(0, 2))) {
+      return multiLineResponse
+    }
+
+    //  create singleline request from multiLine Response
+    val singleLine = verifyAddress.getFormattedAddress + " " + verifyAddress.getCountry.getIsoAlpha3Code
```

Comment:
we need to be careful here (and above) - whenever we use an output as in input.  what if the verify failed? (it can)
We shoudl use the input in that case

Reply:
Ok, I will take care of this